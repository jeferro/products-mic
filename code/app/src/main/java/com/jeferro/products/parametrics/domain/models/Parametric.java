package com.jeferro.products.parametrics.domain.models;

import com.jeferro.products.parametrics.domain.exceptions.ParametricValueNotFoundException;
import com.jeferro.products.parametrics.domain.models.values.ParametricValueId;
import com.jeferro.products.parametrics.domain.models.values.ParametricValues;
import com.jeferro.shared.ddd.domain.models.projection.Projection;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;
import com.jeferro.shared.locale.domain.models.LocalizedField;
import lombok.Getter;

@Getter
public class Parametric extends Projection<ParametricId> {

    private final LocalizedField name;

    private final ParametricValues values;

    public Parametric(
            ParametricId id,
            LocalizedField name,
            ParametricValues values) {
        super(id);

        this.name = name;
        this.values = values;
    }

    public static Parametric createOf(ParametricId id, LocalizedField name, ParametricValues values) {
        ValueValidationUtils.isNotNull(name, "name", Parametric.class);
        ValueValidationUtils.isNotEmpty(values, "values", Parametric.class);

        return new Parametric(id, name, values);
    }

    public boolean validate(ParametricValueId parametricValueId) {
        if (notContainsParametricValue(parametricValueId)) {
            throw ParametricValueNotFoundException.createOf(this, parametricValueId);
        }

        return true;
    }

    private boolean notContainsParametricValue(ParametricValueId parametricValueId) {
        return values.stream()
                .noneMatch(parametricValue -> parametricValue.hasSameId(parametricValueId));
    }
}
