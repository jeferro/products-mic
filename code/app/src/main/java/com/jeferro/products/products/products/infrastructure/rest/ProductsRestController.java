package com.jeferro.products.products.products.infrastructure.rest;

import com.jeferro.products.generated.rest.v1.apis.ProductsApi;
import com.jeferro.products.generated.rest.v1.dtos.CreateProductInputRestDTO;
import com.jeferro.products.generated.rest.v1.dtos.ProductFilterOrderRestDTO;
import com.jeferro.products.generated.rest.v1.dtos.ProductRestDTO;
import com.jeferro.products.generated.rest.v1.dtos.ProductSummaryListRestDTO;
import com.jeferro.products.generated.rest.v1.dtos.UpdateProductInputRestDTO;
import com.jeferro.products.products.products.infrastructure.rest.mappers.ProductRestMapper;
import com.jeferro.shared.ddd.application.bus.UseCaseBus;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequiredArgsConstructor
public class ProductsRestController implements ProductsApi {

    private final ProductRestMapper productRestMapper = ProductRestMapper.INSTANCE;

    private final UseCaseBus useCaseBus;

    @Override
    public ProductSummaryListRestDTO searchProducts(Integer pageNumber, Integer pageSize,
                                                      ProductFilterOrderRestDTO order,
                                                      Boolean ascending,
                                                      String name) {
        var params = productRestMapper.toSearchProductsParams(pageNumber, pageSize, order, ascending, name);

        var products = useCaseBus.execute(params);

        return productRestMapper.toSummaryDTO(products);
    }

    @Override
    public ProductRestDTO createProduct(CreateProductInputRestDTO createProductInputRestDTO) {
        var params = productRestMapper.toCreateProductParams(createProductInputRestDTO);

        var product = useCaseBus.execute(params);

        return productRestMapper.toDTO(product);
    }

    @Override
    public ProductRestDTO getProduct(String productCode) {
        var params = productRestMapper.toGetProductParams(productCode);

        var product = useCaseBus.execute(params);

        return productRestMapper.toDTO(product);
    }

    @Override
    public ProductRestDTO updateProduct(String productCode, UpdateProductInputRestDTO updateProductInputRestDTO) {
        var params = productRestMapper.toUpdateProductParams(productCode, updateProductInputRestDTO);

        var user = useCaseBus.execute(params);

        return productRestMapper.toDTO(user);
    }

    @Override
    public ProductRestDTO publishProduct(String productCode) {
        var params = productRestMapper.toPublishProductParams(productCode);

        var user = useCaseBus.execute(params);

        return productRestMapper.toDTO(user);
    }

    @Override
    public ProductRestDTO unpublishProduct(String productCode) {
        var params = productRestMapper.toUnpublishProductParams(productCode);

        var user = useCaseBus.execute(params);

        return productRestMapper.toDTO(user);
    }

    @Override
    public ProductRestDTO deleteProduct(String productCode) {
        var params = productRestMapper.toDeleteProductParams(productCode);

        var user = useCaseBus.execute(params);

        return productRestMapper.toDTO(user);
    }
}
