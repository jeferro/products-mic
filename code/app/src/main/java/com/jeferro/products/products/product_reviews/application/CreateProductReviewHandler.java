package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.CreateProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.exceptions.ProductReviewAlreadyExistsException;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.models.ProductReviewId;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.products.products.products.domain.models.ProductCode;
import com.jeferro.products.products.products.domain.repositories.ProductsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.events.EventBus;
import com.jeferro.shared.ddd.domain.exceptions.auth.ForbiddenException;
import com.jeferro.shared.ddd.domain.models.auth.Auth;
import com.jeferro.shared.ddd.domain.models.auth.UserAuth;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class CreateProductReviewHandler extends Handler<CreateProductReviewParams, ProductReview> {

    private final ProductsRepository productsRepository;

    private final ProductReviewsRepository productReviewsRepository;

    private final EventBus eventBus;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public ProductReview execute(Context context, CreateProductReviewParams params) {
        Auth auth = context.getAuth();

        var productCode = ensureProductExists(params);

        ensureUserIsAuthenticated(auth);

        ensureProductReviewDoesNotExists(context, productCode);

        return createProductReview(context, params, productCode);
    }

    private void ensureUserIsAuthenticated(Auth auth) {
        if (!(auth instanceof UserAuth)) {
            throw ForbiddenException.createOfNotUserAuth(auth);
        }
    }

    private ProductCode ensureProductExists(CreateProductReviewParams params) {
        var productCode = params.getProductCode();

        var product = productsRepository.findByIdOrError(productCode);

        return product.getCode();
    }

    private void ensureProductReviewDoesNotExists(Context context, ProductCode productCode) {
        var auth = context.getAuth();

        var productReviewId = ProductReviewId.createOf(auth, productCode);

        productReviewsRepository.findById(productReviewId)
                .ifPresent(current -> {
                    throw ProductReviewAlreadyExistsException.createOf(productReviewId);
                });
    }

    private ProductReview createProductReview(Context context, CreateProductReviewParams params, ProductCode productCode) {
        var auth = context.getAuth();
        var locale = context.getLocale();

        var comment = params.getComment();

        var productReview = ProductReview.createOf(productCode, locale, comment, auth);

        productReviewsRepository.save(productReview);

        eventBus.sendAll(productReview);
        return productReview;
    }
}
