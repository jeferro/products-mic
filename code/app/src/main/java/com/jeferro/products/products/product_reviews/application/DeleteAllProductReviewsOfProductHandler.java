package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.DeleteAllProductReviewsOfProductParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.events.EventBus;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.ADMIN;

@Component
@RequiredArgsConstructor
public class DeleteAllProductReviewsOfProductHandler extends Handler<DeleteAllProductReviewsOfProductParams, Void> {

    private final ProductReviewsRepository productReviewsRepository;

    private final EventBus eventBus;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(ADMIN);
    }

    @Override
    public Void execute(Context context, DeleteAllProductReviewsOfProductParams params) {
        var productCode = params.getProductCode();

        var productReviews = productReviewsRepository.findAllByProductCode(productCode);

        if (productReviews.isEmpty()) {
            return null;
        }

        productReviews.forEach(ProductReview::deleteBySystem);

        var productReviewIds = productReviews.getIds();

        productReviewsRepository.deleteAllById(productReviewIds);

        eventBus.sendAll(productReviews);

        return null;
    }
}
