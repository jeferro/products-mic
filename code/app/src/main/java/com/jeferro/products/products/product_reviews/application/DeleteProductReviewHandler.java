package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.DeleteProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.events.EventBus;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class DeleteProductReviewHandler extends Handler<DeleteProductReviewParams, ProductReview> {

    private final ProductReviewsRepository productReviewsRepository;

    private final EventBus eventBus;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public ProductReview execute(Context context, DeleteProductReviewParams params) {
        var productReview = ensureProductReviewExists(params);

        return deleteProductReview(context, productReview);
    }

    private ProductReview ensureProductReviewExists(DeleteProductReviewParams params) {
        var productReviewId = params.getProductReviewId();

        return productReviewsRepository.findByIdOrError(productReviewId);
    }

    private ProductReview deleteProductReview(Context context, ProductReview productReview) {
        var auth = context.getAuth();

        productReview.deleteByUser(auth);

        productReviewsRepository.deleteById(productReview.getId());

        eventBus.sendAll(productReview);

        return productReview;
    }
}
