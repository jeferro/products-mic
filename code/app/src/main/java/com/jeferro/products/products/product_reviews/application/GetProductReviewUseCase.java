package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.GetProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.UseCase;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class GetProductReviewUseCase extends UseCase<GetProductReviewParams, ProductReview> {

    private final ProductReviewsRepository productReviewsRepository;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public ProductReview execute(Context context, GetProductReviewParams params) {
        var productReviewId = params.getProductReviewId();

        return productReviewsRepository.findByIdOrError(productReviewId);
    }
}
