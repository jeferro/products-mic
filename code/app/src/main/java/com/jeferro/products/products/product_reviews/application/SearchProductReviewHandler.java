package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.SearchProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.Handler;
import com.jeferro.shared.ddd.domain.models.aggregates.PaginatedList;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class SearchProductReviewHandler extends Handler<SearchProductReviewParams, PaginatedList<ProductReview>> {

    private final ProductReviewsRepository productReviewsRepository;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public PaginatedList<ProductReview> execute(Context context, SearchProductReviewParams params) {
        var productCode = params.getProductCode();

        return productReviewsRepository.findAllByProductCode(productCode);
    }
}
