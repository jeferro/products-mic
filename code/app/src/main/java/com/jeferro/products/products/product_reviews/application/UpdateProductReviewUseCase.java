package com.jeferro.products.products.product_reviews.application;

import com.jeferro.products.products.product_reviews.application.params.UpdateProductReviewParams;
import com.jeferro.products.products.product_reviews.domain.models.ProductReview;
import com.jeferro.products.products.product_reviews.domain.repositories.ProductReviewsRepository;
import com.jeferro.shared.ddd.application.UseCase;
import com.jeferro.shared.ddd.domain.events.EventBus;
import com.jeferro.shared.ddd.domain.models.context.Context;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Set;

import static com.jeferro.products.shared.application.Roles.USER;

@Component
@RequiredArgsConstructor
public class UpdateProductReviewUseCase extends UseCase<UpdateProductReviewParams, ProductReview> {

    private final ProductReviewsRepository productReviewsRepository;

    private final EventBus eventBus;

    @Override
    public Set<String> getMandatoryUserRoles() {
        return Set.of(USER);
    }

    @Override
    public ProductReview execute(Context context, UpdateProductReviewParams params) {
        var productReview = ensureProductReviewExists(params);

        return updateProductReview(context, params, productReview);
    }

    private ProductReview ensureProductReviewExists(UpdateProductReviewParams params) {
        var productReviewId = params.getProductReviewId();

        return productReviewsRepository.findByIdOrError(productReviewId);
    }

    private ProductReview updateProductReview(Context context, UpdateProductReviewParams params, ProductReview productReview) {
        var auth = context.getAuth();
        var locale = context.getLocale();

        var comment = params.getComment();

        productReview.update(comment, locale, auth);

        productReviewsRepository.save(productReview);

        eventBus.sendAll(productReview);

        return productReview;
    }
}
