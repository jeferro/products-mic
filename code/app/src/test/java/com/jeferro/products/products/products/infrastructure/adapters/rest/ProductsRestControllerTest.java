package com.jeferro.products.products.products.infrastructure.adapters.rest;

import com.jeferro.products.products.products.domain.models.ProductMother;
import com.jeferro.products.products.products.infrastructure.rest.ProductsRestController;
import com.jeferro.products.shared.application.StubUseCaseBus;
import com.jeferro.products.shared.infrastructure.adapters.rest.RestControllerTest;
import com.jeferro.products.shared.infrastructure.adapters.utils.ApprovalUtils;
import com.jeferro.shared.ddd.domain.models.aggregates.PaginatedList;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

@WebMvcTest(ProductsRestController.class)
class ProductsRestControllerTest extends RestControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private StubUseCaseBus stubHandlerBus;

    @Test
    void execute_list_products_on_request() throws Exception {
        var products = PaginatedList.createOfItems(
                ProductMother.apple(),
                ProductMother.pear()
        );
        stubHandlerBus.init(products);

        var requestBuilder = MockMvcRequestBuilders.get("/v1/products?pageNumber=0&pageSize=10&&order=NAME&ascending=true&name=apple")
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_create_product_on_request() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestContent = """
                {
                  "code": "%s",
                  "typeId": "%s",
                  "name": {
                    "en-US": "Apple"
                  }
                }"""
                .formatted(apple.getCode(), apple.getTypeId());

        var requestBuilder = MockMvcRequestBuilders.post("/v1/products")
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN)
                .content(requestContent);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_get_product_on_request() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestBuilder = MockMvcRequestBuilders.get("/v1/products/" + apple.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_update_product_on_request() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestContent = """
                {
                  "name": {
                    "en-US": "%s"
                  }
                }"""
                .formatted(apple.getName());

        var requestBuilder = MockMvcRequestBuilders.put("/v1/products/" + apple.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN)
                .content(requestContent);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void execute_delete_product_on_request() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestBuilder = MockMvcRequestBuilders.delete("/v1/products/" + apple.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.ACCEPT_LANGUAGE, ACCEPT_LANGUAGE_EN)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_USER_TOKEN);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstParamOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

}
