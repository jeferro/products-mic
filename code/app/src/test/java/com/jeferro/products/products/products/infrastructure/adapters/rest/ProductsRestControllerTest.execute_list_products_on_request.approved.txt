data[0] = SearchProductsParams[filter=ProductFilter[name=apple,ascending=true,order=NAME,pageNumber=0,pageSize=10]]
data[1] = 200
data[2] = {"pageNumber":0,"pageSize":2,"totalItems":2,"items":[{"code":"apple","typeId":"fruit","name":{"en-US":"Apple"},"status":"PUBLISHED"},{"code":"pear","typeId":"fruit","name":{"en-US":"Pear"},"status":"PUBLISHED"}]}
